MREUnityRuntime/MREUnityRuntimeLib/Behaviors/Contexts/BehaviorContextFactory.cs

```diff
@@ -0,0 +1,73 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+using MixedRealityExtension.API;
+using MixedRealityExtension.App;
+using MixedRealityExtension.Core.Interfaces;
+using System;
+using System.Collections.Generic;
+
+using FactoryFunc = System.Func<
+	MixedRealityExtension.Core.Interfaces.IActor,
+	System.WeakReference<MixedRealityExtension.App.MixedRealityExtensionApp>,
+	MixedRealityExtension.Behaviors.Contexts.BehaviorContextBase>;
+
+namespace MixedRealityExtension.Behaviors.Contexts
+{
+	internal class BehaviorContextFactory
+	{
+		private static Dictionary<BehaviorType, FactoryFunc> s_factoryMethods =
+			new Dictionary<BehaviorType, FactoryFunc>();
+
+		static BehaviorContextFactory()
+		{
+			s_factoryMethods.Add(BehaviorType.Target, CreateTargetBehaviorContext);
+			s_factoryMethods.Add(BehaviorType.Button, CreateButtonBehaviorContext);
+			s_factoryMethods.Add(BehaviorType.Pen, CreatePenBehaviorContext);
+		}
+
+		internal static BehaviorContextBase CreateContext(BehaviorType behaviorType, IActor actor, WeakReference<MixedRealityExtensionApp> appRef)
+		{
+			if (s_factoryMethods.TryGetValue(behaviorType, out FactoryFunc factoryMethod))
+			{
+				return factoryMethod.Invoke(actor, appRef);
+			}
+
+			MixedRealityExtensionApp app;
+			if (appRef.TryGetTarget(out app))
+			{
+				app.Logger.LogError($"Trying to create a behavior of type {behaviorType.ToString()}, but no handler is registered for the given type.");
+			}
+			return null;
+		}
+
+		private static BehaviorContextBase CreateTargetBehaviorContext(IActor actor, WeakReference<MixedRealityExtensionApp> appRef)
+		{
+			var behaviorFactory = MREAPI.AppsAPI.BehaviorFactory;
+			var context = new TargetBehaviorContext();
+			var targetBehavior = behaviorFactory.GetOrCreateTargetBehavior(actor, context);
+			if (targetBehavior == null) throw new NullReferenceException("Application failed to create a target behavior for the MRE runtime.");
```

Comment:
If this throws, are you certain all codepaths handle it correctly? Will expected reply messages be sent, for example? Wondering if it would be better to log this back to the app instead.

Reply:
We have always thrown exceptions like this in the behavior system if something goes catastrophically wrong, which it would need to in this case.  If we want to discuss changing how we handle errors with regards to the app for the behavior case here and all up, I am open to that discussion.  Though I feel that this is change worthy of its own body of work.

---
In reply to: [428862435](https://github.com/microsoft/mixed-reality-extension-unity/pull/160#discussion_r428862435) [](ancestors = 428862435)